### PR TITLE
Extract out a common shrinking interface

### PR DESCRIPTION
**Pull request by DRMacIver**
As part of #1423 we're going to start to have other sorts of shrinkers around, not just the lexical minimizer. This creates a common shrinking interface that currently only has one implementation but will very soon have more.

NB. the main shrinker in conjecture is not an implementation of this interface. The reason for that is that while these ones shrink concrete values, the main conjecture shrinker operates on conjecture data objects, which are a bit weirder. This could probably still be made to work, but it's not totally obvious how yet.

### CODE REVIEW

**Comment by DRMacIver** [comment]
Ugh. FINALLY. This should be ready to review now.

Once again I'm not sure why this was so painful other than "I suspect a lot of our tests are right on the boundary of failing and minor changes can push them over the edge". 😢

File: hypothesis-python/RELEASE.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+RELEASE_TYPE: patch
+
+This release makes some internal changes to how shrinking is structure. It is
```
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst]: Nit: "shrinking is structure[d]", 'mostly a pure refactoring' is vague.  I'd rephrase this along the lines of

> This release makes some internal changes which facilitate further improvements to the shrinking engine.  You may see some changes in the final shrunk examples, but this is unlikely to be significant.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinking/integer.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    """
+
+    def short_circuit(self):
+        for i in hrange(3):
+            if self.consider(i):
+                return True
+        return False
+
+    def check_invariants(self, value):
+        assert value >= 0
```
[the remark this answers is not in the document]
Comment (DRMacIver) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinking/integer.py]: No, I don't think this is a problem. These are actual assertions in the sense of "it is a bug if this expression is ever false", they're not intended as argument validation per se they're just convenient hooks.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinking/integer.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    def left_is_better(self, left, right):
+        return left < right
+
+    def run_step(self):
+        assert self.current > 2
+        self.consider(self.current - 2)
+        self.consider(self.current - 1)
+        assert self.current > 2
```
[the remark this answers is not in the document]
Comment (DRMacIver) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinking/integer.py]: Same answer - these should be impossible to hit and in this case so obviously are (because we have the `value in self.seen` check and we've already tried 0, 1, 2 in the short circuit path) are more of a comment than a real error check.

These are intended to be lightweight things to make debugging problems easier. The alternative isn't that I rewrite them to be good error messages, it's that I remove them altogether because they're not worth the effort.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinking/integer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinking/integer.py]: Fair enough!  I did have "explain to Zac" in mind as a way of addressing them, hence the approval regardless.  But better to check and learn something than miss a problem :smile:
